bcda/auth/logger.go

```diff
@@ -0,0 +1,98 @@
+package auth
+
+import (
+	"os"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+var logger *logrus.Logger
+
+const (
+	// tknExpired       = "AccessTokenExpired"
+	// tknInvalid       = "AccessTokenInvalid"
+	// tknIssued        = "AccessTokenIssued"
+	// tknRevoked       = "AccessTokenRevoked"
+	// credsIssued      = "ClientCredentialsIssued"
+	// credsRevoked     = "ClientCredentialsRevoked"
+	// reqAuthorized    = "RequestAuthorized"
+	// reqNotAuthorized = "RequestNotAuthorized"
+	// requesterUnknown = "RequesterUnknown"
+	// secretCreated    = "SecretCreated"
+	// secretFast       = "SecretTooFast"
+	// secretSlow       = "SecretTooSlow"
+	// a failure of something other than auth logic: duplicate db entries, network failures
+	// sysFailure = "SystemFailure"
+	// operation events
+	opFailed    = "OperationFailed"
+	opStarted   = "OperationStarted"
+	opSucceeded = "OperationSucceeded"
+	// service events
+	// halted  = "ServiceHalted"
+	// started = "ServiceStarted"
+	// stopped = "ServiceStopped"
+)
+
+type event struct {
+	clientID   string
+	elapsed    int
+	help       string
+	op         string
+	tokenID    string
+	trackingID string
+}
+
+func init() {
+	logger = logrus.New()
+	logger.Formatter = &logrus.JSONFormatter{}
+	logger.Formatter.(*logrus.JSONFormatter).TimestampFormat = time.RFC3339Nano
+
+	filePath, success := os.LookupEnv("AUTH_LOG")
+	if success {
+		/* #nosec -- 0640 permissions required for Splunk ingestion */
+		file, err := os.OpenFile(filePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0640)
+
+		if err == nil {
+			logger.SetOutput(file)
+		} else {
+			logger.Info("Failed to open Auth log file; using default stderr")
+		}
+	} else {
+		logger.Info("No Auth log location provided; using default stderr")
+	}
+}
+
+func mergeNonEmpty(data event) *logrus.Entry {
```

Comment:
Based on my experience, I think you avoided future bugs by writing this method.  🐛 🔨 💯